storage docs (#2045)
fetchai/agents-aea: docs/generic-storage.md

The AEA generic storage: description and usage.

## AEA Generic Storage
AEA generic storage is system allows AEA skill's components to store data permanently and use it any time.
The primary scenario: to save AEA data on shutdown and load back on startup.
Storage provides API for general data manipulation in key, object style.

## Configuration
Storage enabled by providing agent configuration option named `storage_uri`
Storage uri consists of backend name and string data provided to selected backend.

Storage uri structure is `<BACKEND_NAME>://[Optional string]`
Example: `storage_uri: sqlite://./some_file.db` tells to use sqlite backend and store data in ./some_file.db

Supported backends:
* sqlite - bundled with python simple sql engine that uses file or in-memory storage.

## Dialogues and Storage integration

> One of the most useful cases is integration of dialogues and storage, that helps easily keep dialogues state during the agent restart.
> Another benefit is memory saving with offloading feature.

Reviewer: ```suggestion
One of the most useful cases is the integration of the dialogues subsystem and storage. It helps maintain dialogues state during agent restarts and reduced memory requirements due to the offloading feature.
```

### Keep terminal state dialogues

> Dialogues/Model skill has the optional boolean argument `keep_terminal_state_dialogues`
> specifies should dialogue reached terminal state kept in memory or not.
> if `keep_terminal_state_dialogues` is False, dialogues that reaches  terminal state  is removed from the memory and can not be used anymore
> if `keep_terminal_state_dialogues` is True, dialogues that reaches  terminal state  is removed from the memory and can not be used anymore
> 
> It useful to save memory with dialogues that are in terminal state and probably will be never used again.
> 
> Default behaviour on keep terminals state dialogues is set according to the protocol specification but can be set explicitly with skill configuration section.
> 
> 
> Skill configuration to keep terminated dialogues for DefaultDialogues.
> Example:
> ```
> models:
>   default_dialogues:
>     args:
>     	keep_terminal_state_dialogues: true
>     class_name: DefaultDialogues
> ```

Reviewer: ```suggestion
The Dialogues class has the optional boolean argument `keep_terminal_state_dialogues`
which specifies whether a dialogue which has reached its terminal state is kept in memory or not. If `keep_terminal_state_dialogues` is `False`, dialogues that reach a terminal state are removed from memory and can not be used anymore. If `keep_terminal_state_dialogues` is `True`, dialogues that reach a terminal state are kept.

It useful to save memory with dialogues that are in terminal state and probably will be never used again.

Default behaviour on keep terminals state dialogues is set according to the protocol specification but can be set explicitly with skill configuration section.

Skill configuration to keep terminated dialogues for DefaultDialogues.
Example:
```
models:
  default_dialogues:
    args:
    	keep_terminal_state_dialogues: true
    class_name: DefaultDialogues
```
```

### Dialogues dump/restore on agent restart

> if storage is enabled: all the dialogues present in memory will be stored on agent's teardown and loaded on agent's start.

Reviewer: ```suggestion
If storage is enabled then all the dialogues present in memory will be stored on agent's teardown and loaded on agent's start.
```

### Offload terminal state dialogues

> If keep options is set and storage is available dialogues in terminal state will be dumped to generic storage and removed from memory.
> This option helps to save memory and handle terminated dialogues with the same functionality as they just kept in memory.
> 
> All the active dialogues will be stored and loaded during agent restart.
> All the terminated offloaded dialogues will stay on agent restart.
> 
> To enable dialogues offloading `keep_terminal_state_dialogues` has to be enabled and storage configured.

Reviewer: ```suggestion
If keep options is set and storage is available dialogues in terminal state will be dumped to generic storage and removed from memory. This option helps to save memory and handle terminated dialogues with the same functionality as when they are kept in memory.

All the active dialogues will be stored and loaded during agent restart. All the terminated offloaded dialogues will stay in storage on agent restart.

To enable dialogues offloading `keep_terminal_state_dialogues` has to be enabled and storage configured.
```

## Manual usage with skill components
Handlers, Behaviours and Models are able to use storage if enabled.

Storage is available with skill context: `self.context.storage`
if `self.context.storage` is not None, storage is enabled and ready to use.

> Generic storage consists of two parts:
> Objects and collections.
> Object consists f  object_id (unique string) and object body. Object body is any json friendly python data type: list, dict, int, float, string ,bool.
> 
> 
> Collection is a group of the objects, objects data types can vary in the same collection.
> Collection name is name consists of letters, numbers and _.

Reviewer: ```suggestion
Generic storage consists of two parts: objects and collections.
Objects consist of the object_id (unique string) and object body. The object body is any json friendly python data type: list, dict, int, float, string, bool.

Collection is a group of the objects, objects data types can vary in the same collection.
Collection name is name consists of letters, numbers and _.
```
